### ISSUE TITLE
Preventing the attacker from redirecting the control flow during runtime if an attacker knows the details of our data-leaking-protection method

### ISSUE THREAD

**Issue by StanPlatinum** [StanPlatinum/elf-respect#7]
@heartever's words reminds me that an attacker may commit a control flow hijack to bypass our data-leakage checking. Our tentative solution is to use Intel TSX (HLE) feature that makes our checker (code for data-leakage check) a transaction. I think the idea is great and it's easy to implement. However, it seems that the introduced performance overhead is not negligible...

**Comment by mxz297** [comment]
I started to read the SGX-shield paper. My first impression is that I feel control flow hijack is an orthogonal problem:

In our data leakage threat model: the client is benign and providing data to untrusted application provider. The client is worrying that the untrusted application provider may contain data leakage code that can leak client's important data on untrusted running platform.

The control flow hijack threat model: the client is malicious and is trying to provided malicious input to trusted application provider.

I think that control flow hijack is out of scope for this project.

**Comment by StanPlatinum** [comment]
After discussing with @mxz297  xiaozhu and @heartever 师兄, I realize that what we really need is not CFI but a defense against attackers executing some malicious code on a writable page (inside/outside an enclave).

So I guess Gruss's 'sgx-rop' paper would tell us something about it. The link is attached on another issue. You can check it.

**Comment by StanPlatinum** [comment]
I also wrote our threat model in my mind down to our project wiki. Pls have a look at it too, thanks!

**Comment by heartever** [comment]
@lpjlwj We should still need (partial CFI) to make sure address checking cannot be passed by redirecting the control flow. If the code pointer is redirected to the address after our checking is finished, then it is possible to leak the secret without being noticed.

Another important attack strategy is that the attacker makes uses of the existence of W+X page to construct leaking code during runtime. We can prevent this if we put all W+X page in the lowest addresses and make sure every memory write only targets higher addresses (SGX-Shield paper: Software DEP section).

In summary, we need to check memory write operations to make sure they do not write to addresses
* Outside the enclave or
* In a W+X page inside the enclave.

Actually we do not "check" the address, but mask the address to make it in a region that does not leak secret (without affecting the functionalities).

**Comment by mxz297** [comment]
> @lpjlwj We should still need (partial CFI) to make sure address checking cannot be passed by redirecting the control flow. If the code pointer is redirected to the address after our checking is finished, then it is possible to leak the secret without being noticed.

I do not quite understand this. Since we will perform binary rewriting to add checks after the application binary is already compiled and generated, the attacker does not have the chance to re-install control flow transfer to avoid our instrumentation.  

> 
> Another important attack strategy is that the attacker makes uses of the existence of W+X page to construct leaking code during runtime. We can prevent this if we put all W+X page in the lowest addresses and make sure every memory write only targets higher addresses (SGX-Shield paper: Software DEP section).

Are there any real applications running in SGX that uses JIT in SGX? 

If the benign application is using JIT In SGX, then it will be broken under this design, because the design does not allow writing to W+X page.

If no benign application is using JIT in SGX, then we should try to not allow creating W+X page. I know that outside SGX, you can call `mprotect` to change page permission. How does a person change memory permission in SGX?

**Comment by heartever** [comment]
> > @lpjlwj We should still need (partial CFI) to make sure address checking cannot be passed by redirecting the control flow. If the code pointer is redirected to the address after our checking is finished, then it is possible to leak the secret without being noticed.
> 
> I do not quite understand this. Since we will perform binary rewriting to add checks after the application binary is already compiled and generated, the attacker does not have the chance to re-install control flow transfer to avoid our instrumentation.
> 

The attacker can prepare some gadget in the binary. After instrument the binary looks like
```
0  mov 0x3ff000, %rax
1  ... ; address checking code
2  ... ; address checking code
3  mov $secret, [%rax]
4  ret
```
Then during runtime something such as 
```
mov 0x7fffff00, %rax ; an outer address
call $3
``` 
will bypass our checking.

> > Another important attack strategy is that the attacker makes uses of the existence of W+X page to construct leaking code during runtime. We can prevent this if we put all W+X page in the lowest addresses and make sure every memory write only targets higher addresses (SGX-Shield paper: Software DEP section).
> 
> Are there any real applications running in SGX that uses JIT in SGX?
> 
> If the benign application is using JIT In SGX, then it will be broken under this design, because the design does not allow writing to W+X page.
> 
> If no benign application is using JIT in SGX, then we should try to not allow creating W+X page. I know that outside SGX, you can call `mprotect` to change page permission. How does a person change memory permission in SGX0?

W+X page is necessary to dynamically lode the binary code.

In SGXv2, there are instructions to faithfully change the page properties with the help of untrusted OS. In SGXv1, we need to prevent the writing to W+X pages.

**Comment by mxz297** [comment]
> ```
> 0  mov 0x3ff000, %rax
> 1  ... ; address checking code
> 2  ... ; address checking code
> 3  mov $secret, [%rax]
> 4  ret
> ```
> Then during runtime something such as
> 
> ```
> mov 0x7fffff00, %rax ; an outer address
> call $3
> ```
> will bypass our checking.
> 

I now understand what you mean. Let me expand the discussion. At the binary level, there are direct control flow transfer and indirect control flow transfer. 

Direct control flow transfer such as direct calls and direct jumps are encoded as a PC-relative fashion (PC+offset). When Dyninst performs instrumentation, it makes sure the semantics of the instrumentation. For example, if we want to instrument memory access, the memory instrumentation should be always executed before the memory access itself. So, Dyninst will adjust all direct control flow transfer to make sure instrumentations are never skipped. 

For indirect control flow (indirect calls, indirect jumps, and rets), Dyninst cannot really pre-determine where it goes, so have to rely on a few assumptions. For example, indirect calls are assumed to go to a function entry. Dyninst will then install branches at function entry to make sure it does not skip instrumentation.

Obviously, Dyninst's assumption about indirect control flow does not fit our scenario here. The root cause here is that we are instrumenting untrusted software. In theory, the attack can embed certain indirect control flow calculation so that the indirect CF will skip our instrumentation. I have not seen any real work on this topic, and I feel this is a very interesting separate topic. In practice, because our instrumentation tool is public, the attacker can use the instrumentation tool to first see the instrumentation layout, and then change the code to skip instrumentation.  Therefore, we only need to worry about indirect control flow.

To address this problem, we only need to instrument every indirect control flow sites, and check if the control flow target will skip instrumentation. I do not think TSX is necessary here. The key here is to determine what type of control flow targets should be disallowed. 

> In SGXv2, there are instructions to faithfully change the page properties with the help of untrusted OS. In SGXv1, we need to prevent the writing to W+X pages.

I read a few pages of the SGX shield paper, and understand why W+X pages exist in SGXv1. 

My understanding is that we will need to distinguish memory writes to W+X pages due to dynamic loading from memory writes to W+X pages due to JIT. We will allow the first type of writes, but disallow second type of writes.
